arwen/host/execution.go

```diff
@@ -0,0 +1,376 @@
+package host
+
+import (
+	"fmt"
+
+	"github.com/ElrondNetwork/arwen-wasm-vm/arwen"
+	"github.com/ElrondNetwork/arwen-wasm-vm/wasmer"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+func (host *vmHost) doRunSmartContractCreate(input *vmcommon.ContractCreateInput) (*vmcommon.VMOutput, error) {
+	host.InitState()
+
+	blockchain := host.Blockchain()
+	runtime := host.Runtime()
+	metering := host.Metering()
+	output := host.Output()
+
+	var err error
+	var returnCode vmcommon.ReturnCode
+	var vmOutput *vmcommon.VMOutput
+	defer func() {
+		if err != nil {
+			vmOutput = output.CreateVMOutputInCaseOfError(returnCode, err.Error())
+			fmt.Printf("Arwen error: %s - %s\n", returnCode, err.Error())
+		} else if returnCode != vmcommon.Ok {
+			vmOutput = output.CreateVMOutputInCaseOfError(returnCode, output.ReturnMessage())
+			fmt.Printf("Arwen error: %s - %s\n", returnCode, returnCode)
+		}
+	}()
+
+	address, err := blockchain.NewAddress(input.CallerAddr)
+	if err != nil {
+		returnCode = vmcommon.ExecutionFailed
+		return vmOutput, nil
+	}
+
+	runtime.SetVMInput(&input.VMInput)
+	runtime.SetSCAddress(address)
+	output.AddTxValueToAccount(address, input.CallValue)
+
+	gasForDeployment, err := metering.DeductInitialGasForDirectDeployment(input)
+	if err != nil {
+		returnCode = vmcommon.OutOfGas
+		return vmOutput, nil
+	}
+
+	vmInput := runtime.GetVMInput()
+	vmInput.GasProvided = gasForDeployment
+	err = runtime.CreateWasmerInstance(input.ContractCode, vmInput.GasProvided)
+	if err != nil {
+		returnCode = vmcommon.ContractInvalid
+		return vmOutput, nil
+	}
+
+	idContext := arwen.AddHostContext(host)
+	runtime.SetInstanceContextId(idContext)
+	defer func() {
+		runtime.CleanInstance()
+		arwen.RemoveHostContext(idContext)
+	}()
+
+	result, err := host.callInitFunction()
+	if err != nil {
+		returnCode = vmcommon.FunctionWrongSignature
+		return vmOutput, nil
+	}
+
+	output.DeployCode(address, input.ContractCode)
+	vmOutput = output.CreateVMOutput(result)
+
+	return vmOutput, nil
+}
+
+func (host *vmHost) doRunSmartContractCall(input *vmcommon.ContractCallInput) (*vmcommon.VMOutput, error) {
+	host.InitState()
+	runtime := host.Runtime()
+	output := host.Output()
+	metering := host.Metering()
+	blockchain := host.Blockchain()
+
+	var err error
+	var vmOutput *vmcommon.VMOutput
+	var returnCode vmcommon.ReturnCode
+	defer func() {
+		if err != nil {
+			vmOutput = output.CreateVMOutputInCaseOfError(returnCode, err.Error())
```

Comment:
no need for 2 cases, first is enough. you could call with err.Error + output.ReturnMessage()

Reply:
Done.